tools/register_users_from_csv.py

```diff
@@ -0,0 +1,138 @@
+            "Authorization": f"Bearer {client_token}",
+            "Content-Type": "application/json",
+        }
+        session.headers.update(headers)
+        return session
+
+    def create_user(self, username, first_name, last_name, email):
+        """Create a user."""
+        url = f"{self.scim_url}/Users/{username}"
+        r = self.session.get(url)
+        if r.status_code == 200:
+            print(f"User {username} exists.")
+        if r.status_code == 404:
+            print(f"Creating user {username}.")
+            payload = {
+                "schemas": ["urn:ietf:params:scim:schemas:core:2.0:User"],
+                "userName": username,
+                "name": {
+                    "givenName": first_name,
+                    "familyName": last_name,
+                },
+                "emails": [
+                    {
+                        "value": email,
+                        "type": "work",
+                        "primary": True,
+                    }
+                ]
+            }
+            r = self.session.post(
+                f"{self.scim_url}/Users",
+                json=payload
+            )
+            if r.status_code == 201:
+                print(f"Created user {username}.")
+
+    def add_user_to_group(self, username, allocation):
+        r = self.session.get(
+            f"{self.scim_url}/Groups/{allocation}"
+        )
+        if r.status_code == 200 and username not in [
+            x.get("value") for x in r.json().get("members", {})
+        ]:
+            payload = {
+                "schemas": [
+                    "urn:ietf:params:scim:api:messages:2.0:PatchOp"
+                ],
+                "Operations": [
+                    {
+                        "op": "add",
+                        "value": {
+                            "members": [
+                                {
+                                    "value": username
+                                }
+                            ]
+                        }
+                    }
+                ]
+            }
+            r = self.session.patch(
+                f"{self.scim_url}/Groups/{allocation}",
+                json=payload
+            )
+            if r.status_code in [200, 201]:
+                print(f"Added user {username} to allocation {allocation}.")
+
+
+def main():
+    client = ScimClient(
+        sys.argv[2],
+        "https://keycloak.mss.mghpcc.org",
+        os.environ.get('CLIENT_ID'),
+        os.environ.get('CLIENT_SECRET')
+    )
+
+    with open(sys.argv[1], newline='') as csvfile:
+        reader = csv.reader(csvfile, delimiter=',', quotechar='|')
+        for row in reader:
+            print(f"Processing: {row}.")
+
+            client.create_user(username=row[0],
+                               first_name=row[1],
+                               last_name=row[2],
+                               email=row[0])
+            client.add_user_to_group(username=row[0],
```

Comment:
I wonder if users have to do email validation before they can be added to a group?

Reply:
How is this different from the "Register for an Account" process that they normally would do to request an account from scratch?  We have just taken care of the step where the PI has to add them to the project, but they still have to request an account and wait for manual account approval by NERC before they can use it?  This isn't going to work if they need to log in on the first day of class.

Reply:
@hpdempsey There is no manual account approval. All they have to do is register through RegApp and verify their email address (this creates the Keycloak account). This script precreates the OpenShift user that will correspond to their Keycloak identity. They can log in immediately after that as the permissions have already been set up.

Reply:
The flow is
1. PI gets an allocation approved.
2. PI gets list of users.
3. NERC Admin runs script with list of users.
4. User registers in RegApp, and verify email address.
5. User authenticated to OpenShift and they already have access to namespace.

Reply:
@knikolla This makes sense. I want to confirm the sub steps for step 4, please let me know if this looks right.

4. PI lets the class know they need to register on regapp with their email address as their username
   1. Click [here](https://regapp.mss.mghpcc.org/) and then click Register for an Account
   1. Read and follow the 6 steps. pay special attention to step 5 where you want to ensure you do not change your username from the default of your email address as that is how your course login was created.
   1. As the steps say in step 5 make sure to confirm your address by checking your email and confirming via the link in the validation email
   
@hpdempsey I believe this is going to be the best way to automate the creation of their accounts and still ensure they accept the EULA. Please let us know if you feel I am missing something and if we need to automate the regapp piece of this.

Reply:
It's OK as long as there is no manual step on the NERC side after the bulk upload.   Keep in mind that Red Hat folks use Google for access, so I've never actually seen the steps for someone who uses a BU address.  Why we ask questions....    :-)

Reply:
Thank you for asking questions. Best way to confirm we know what we are talking about.